### PR TITLE
[3.3.5] Pooling system rewrite

### PR DESCRIPTION
**Pull request by Treeston**
## Guess who's back, baby.

So, spawn pooling. You take a bunch of spawns, and you put them in a pool. And then you tell the pool to spawn `x` spawns, and it spawns `x` random spawns from the pool based on the chances specified for each of them. And everything plays nice if grids get loaded and unloaded and stuff, and if one spawn in the pool goes away then when its respawn timer expires another random spawn from the pool is spawned. And you can also put pools in other pools, so you can have a pool that decides which of its sub-pools to spawn, and then each sub-pool picks something to spawn from itself. Gathering nodes do that one.

Anyway. I tore the whole thing out, since it was built around old persistent entities (like, old _old_ entities, pre-dynspawn old), and remade the whole thing since we have this fancy concept of a respawn timer that's not tied to any object now and stuff, which makes stuff _way_ easier by comparison. The old system had to hack its way around a lot of things (and was probably broken in many small and creative ways, if we're being entirely honest).

---

# Changelist

* New `SPAWN_TYPE_POOL (2)`.
  * Multiple respawn timers of the same spawnId are legal now.
* Split quest pooling logic from spawn pooling logic. They both use `pool_template`, though I might split it off too before this is over.
  * Standardized weekly and monthly respawns to also happen at the daily reset hour while I was at it
* Pools can no longer be cross-map. Thread safety and all that.
  * Pools are now per-map. You can have pools inside instances if you feel like it.
* Tons of on-load sanity checking of pool spawns. All those pool fixes pushed to 3.3.5 recently? Yeah.
  * Did we know our Alterac Mountains pool contained nodes in the Barrens? It did!
  * Our Duskwood pool had some nodes in Desolace!
  * _(we call them happy little features)_
* Some cleanups to pooling chance specification
  * Chances no longer need to add to `100.0` - weighting is relative ( `2.0` is twice as likely as `1.0`)
  * No more implicit equal chancing - specify explicit chances for everything (there's a conversion script)
* Merged all respawn timers into `respawn` (from old `creature_respawn` and `gameobject_respawn`) [374597c]
* Merged all spawn pool members into `pool_members` (from old `pool_creature`, `pool_gameobject` and `pool_pool`) [45bc91c]

---

Apart from those, everything _should_ work the same.

You know the drill, it probably doesn't. Go break things.

# To do:
- [ ] database update queries still assume one timer per spawnid, needs fix

### CODE REVIEW

File: src/server/game/Pools/PoolMgr.h [hunk 1]
```diff
@@ -16,197 +16,82 @@
...
+            explicit operator bool() const { return !!parentId; }
+            uint32 parentId;
+            uint32 index;
+        };
 
-        void Spawn1Object(PoolObject* obj);
-        void ReSpawn1Object(PoolObject* obj);
-        void RemoveOneRelation(uint32 child_pool_id);
-        uint32 GetFirstEqualChancedObjectId()
+        static PoolMgr* instance()
```
Comment (Shauren) [inline src/server/game/Pools/PoolMgr.h]: must be in cpp

File: src/server/game/Events/GameEventMgr.cpp [hunk 2]
```diff
@@ -1316,8 +1314,8 @@ void GameEventMgr::GameEventSpawn(int16 event_id)
         return;
     }
 
-    for (IdList::iterator itr = mGameEventPoolIds[internal_event_id].begin(); itr != mGameEventPoolIds[internal_event_id].end(); ++itr)
-        sPoolMgr->SpawnPool(*itr);
+    /*for (IdList::iterator itr = mGameEventPoolIds[internal_event_id].begin(); itr != mGameEventPoolIds[internal_event_id].end(); ++itr)
+        sPoolMgr->SpawnPool(*itr);*/ // @todo
```
Comment (jackpoz) [inline src/server/game/Events/GameEventMgr.cpp]: still some todo's left before merging the PR ?

**Comment by jackpoz** [comment]
> * Standardized weekly and monthly respawns to also happen at the daily reset hour while I was at it

I have a PR open at https://github.com/TrinityCore/TrinityCore/pull/23540 that fixes timezone issues with most of "do this at this hour of the day", maybe I could merge that before this PR so you can use the time-related methods added there
```cpp
time_t LocalTimeHourToUnixTime(time_t day, uint32 hour, bool mustBeAfterDay);
```

**Comment by Jildor** [comment]
#16076 this issue should be fixed after this PR or nothing change? (Only for tag it ;) )

**Comment by Treeston** [comment]
Yes, should be fixed.

**Comment by Killyana** [comment]
There some missing data with the command list respawn:

For non  pooled spawns:
Listing Creature respawns for Mille pointes (zone 400)
SpawnID | Entry | GridXY| Zone | Respawn time (Full)
21297 | 4130 | [19,25] | Mille pointes (400) | 4m57s

For pooled spawn
Listing Pool respawns for Mille pointes (zone 400)
SpawnID | Entry | GridXY| Zone | Respawn time (Full)
32492 | 4133 | [00,00] | <unknown zone> (0) | 2m49s (inactive)
32494 | 4131 | [00,00] | <unknown zone> (0) | 2m51s (inactive)
32493 | 4133 | [00,00] | <unknown zone> (0) | 4m15s (inactive)

**Comment by Treeston** [comment]
Intentional - you can't assign a grid/zone to a pool.

**Comment by Killyana** [comment]
we can maybe replace GridXY| Zone by the poolID|comment

**Comment by Treeston** [comment]
spawnid is old poolid, and the core doesn't load the comments

**Comment by jackpoz** [comment]
>/home/circleci/project/src/server/game/Events/GameEventMgr.cpp:950:107: fatal error: format specifies type 'unsigned int' but the argument has type 'std::unordered_multimap<short, unsigned int, std::hash<short>, std::equal_to<short>, std::allocator<std::pair<const short, unsigned int> > >::size_type' (aka 'unsigned long') [-Wformat]
            TC_LOG_INFO("server.loading", ">> Loaded %u spawn group relations for game events in %u ms.", mGameEventSpawnGroups.size(), GetMSTimeDiffToNow(oldMSTime));

**Comment by Treeston** [comment]
work resuming now that #23627 is merged

some todos to keep you posted:
* database update queries still assume one timer per spawnid, needs fix
* split off some of the necessary refactors and merge them into 335 separately to reduce diff size

**Comment by jackpoz** [comment]
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

**Comment by al3xc1985** [comment]
@Treeston still working on this?

**Pull request closed without merging**